Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp

```diff
@@ -115,8 +117,54 @@ void MappingButton::Detect()
     // Avoid that the button press itself is registered as an event
     Common::SleepCurrentThread(100);
 
-    const auto expr = MappingCommon::DetectExpression(
-        m_reference, dev.get(), m_parent->GetController()->GetDefaultDevice());
+    std::vector<std::future<QPair<QString, QString>>> futures;
+
+    QString expr;
+
+    if (m_parent->GetParent()->IsMappingAllDevices())
+    {
+      for (const std::string& device_str : g_controller_interface.GetAllDeviceStrings())
+      {
+        ciface::Core::DeviceQualifier devq;
+        devq.FromString(device_str);
+
+        auto adev = g_controller_interface.FindDevice(devq);
+
+        auto future = std::async(std::launch::async, [this, devq, adev, device_str] {
+          return QPair<QString, QString>(
+              QString::fromStdString(device_str),
+              MappingCommon::DetectExpression(m_reference, adev.get(),
+                                              m_parent->GetController()->GetDefaultDevice()));
+        });
+
+        futures.push_back(std::move(future));
+      }
+
+      bool done = false;
+
+      while (!done)
+      {
+        for (auto& future : futures)
+        {
+          const auto status = future.wait_for(std::chrono::milliseconds(10));
+          if (status == std::future_status::ready)
+          {
+            const auto pair = future.get();
+            expr = QStringLiteral("`%1:%2`")
```

Comment:
Is this supposed to assign, rather than concatenate?

Reply:
Could you further elaborate as to what you mean by that?

Reply:
Ah, disregard that one, I misread the ending part of the loop. I missed the `break;`  (so I thought it was a logical error of assigning when concatenating was intended, since it would just spin on assignments), my bad.